public/app/features/dashboard-scene/settings/AnnotationsEditView.tsx

```diff
@@ -36,58 +37,130 @@ export class AnnotationsEditView extends SceneObjectBase<AnnotationsEditViewStat
     return getDashboardSceneFor(this);
   }
 
-  private get _dataLayers(): SceneDataLayerProvider[] {
-    return sceneGraph.getDataLayers(this._dashboard);
-  }
-
-  public getSceneDataLayers(): SceneDataLayers | undefined {
+  public getSceneDataLayers(): SceneDataLayers {
     const data = sceneGraph.getData(this);
 
     if (!(data instanceof SceneDataLayers)) {
-      return undefined;
+      throw new Error('SceneDataLayers not found');
     }
 
     return data;
   }
 
   public getAnnotationsLength(): number {
-    return this._dataLayers.filter((layer) => layer.topic === DataTopic.Annotations).length;
+    return this.getSceneDataLayers().state.layers.filter((layer) => layer.topic === DataTopic.Annotations).length;
   }
 
   public getDashboard(): DashboardScene {
     return this._dashboard;
   }
 
   public onNew = () => {
-    console.log('todo: onNew');
+    const newAnnotationQuery: AnnotationQuery = {
+      name: newAnnotationName,
+      enable: true,
+      datasource: getDataSourceRef(getDataSourceSrv().getInstanceSettings(null)!),
+      iconColor: 'red',
+    };
+
+    const newAnnotation = new DashboardAnnotationsDataLayer({
+      key: `annotations-${newAnnotationQuery.name}`,
+      query: newAnnotationQuery,
+      name: newAnnotationQuery.name,
+      isEnabled: Boolean(newAnnotationQuery.enable),
+      isHidden: Boolean(newAnnotationQuery.hide),
+    });
+
+    const data = this.getSceneDataLayers();
+
+    const layers = [...data.state.layers];
+    newAnnotation.activate();
+
+    //keep annotation layers together
+    layers.splice(this.getAnnotationsLength(), 0, newAnnotation);
+
+    data.setState({
+      layers,
+    });
+
+    this.setState({ editIndex: this.getAnnotationsLength() - 1 });
```

Comment:
I don't fully understand why we need this, why don't we do something like this:
```suggestion
    const newAnnotation = new DashboardAnnotationsDataLayer({
      key: `annotations-${newAnnotationQuery.name}`,
      query: newAnnotationQuery,
      name: newAnnotationQuery.name,
      isEnabled: Boolean(newAnnotationQuery.enable),
      isHidden: Boolean(newAnnotationQuery.hide),
    });
    newAnnotation.activate();

    const data = this.getSceneDataLayers();

    const layers = [...data.state.layers, newAnnotation];

    data.setState({
      layers,
    });

    this.setState({ editIndex: this.getAnnotationsLength() - 1 });
```

Not fully sure, but shouldn't we activate it once the object is in the hierarchy? just in case the activation handler needs anything from the parent.

Reply:
I'm not sure I follow, are you suggesting activating the new annotation after it's set in the layers? Also the splice is there to keep annotations grouped since there might be other data layer types in between.

Reply:
Sorry Victor, let me explain 🙇 

> I'm not sure I follow, are you suggesting activating the new annotation after it's set in the layers?

For the scene objects, we usually have activation handlers. They may access the parents or the dashboard scene. Activating the scene object not being part of the dashboard scene yet could cause an error as the dashboard scene won't be found. Probably, in this case, the handler doesn't need anything from the parent, but I'd consider it as a good practice.

> Also the splice is there to keep annotations grouped since there might be other data layer types in between.

When you say grouped, what do you mean? I don't understand why it is different from creating a new array and appending the new annotation in the end or even using `push()`. 
AFAIK, if I have `a = [annot1, annot2]` and I do `[...a, annot3]`, the output is `[annot1, annot2, annot3]` same as using `splice()` or `push()`. 
I mean, it works, just I'm curious about it since it is not a very common way to add new methods when you create a new array anyway 😄